include/circt/Dialect/Moore/MooreTypes.td

```diff
@@ -0,0 +1,44 @@
+//===- MooreTypes.td - Moore type definitions --------------*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// SystemVerilog types for the Moore dialect. Includable in other dialects.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_MOORE_MOORETYPES
+#define CIRCT_DIALECT_MOORE_MOORETYPES
+
+include "circt/Dialect/Moore/MooreDialect.td"
+
+//===----------------------------------------------------------------------===//
+// LValue / RValue wrapper types
+//===----------------------------------------------------------------------===//
+
+class MooreLValueTypeOf<list<Type> allowedTypes>
+  : ContainerType<AnyTypeOf<allowedTypes>,
+    CPred<"$_self.isa<moore::LValueType>()">,
+    "$_self.cast<moore::LValueType>().getNestedType()", "Moore LValue type">;
+
+class MooreRValueTypeOf<list<Type> allowedTypes>
+  : ContainerType<AnyTypeOf<allowedTypes>,
+    CPred<"$_self.isa<moore::RValueType>()">,
+    "$_self.cast<moore::RValueType>().getNestedType()", "Moore RValue type">;
+
+
+def MooreLValueType : MooreLValueTypeOf<[AnyType]>;
+def MooreRValueType : MooreRValueTypeOf<[AnyType]>;
+
+//===----------------------------------------------------------------------===//
+// Integer atom types
+//===----------------------------------------------------------------------===//
+
+def MooreIntType : DialectType<MooreDialect,
```

Comment:
I love that you're using `int` as a stand-in for future more elaborate type system work. Pretty sure that SV type system is going to need some love to bring it to an MLIR dialect. Focusing on MIR and the lowering is very nice as a starting point!

Reply:
Yes this simple int type will probably be removed in the future. But implementing the whole packed type hierarchy in this initial commit would have been too much just to get a simple example running.